ai/movie-recommendation/src/main/java/com/hazelcast/samples/ai/movies/TextSimilaritySearchImap.java

```diff
@@ -0,0 +1,279 @@
+import com.hazelcast.query.Predicates;
+import dev.langchain4j.model.embedding.onnx.allminilml6v2.AllMiniLmL6V2EmbeddingModel;
+import dev.langchain4j.model.embedding.EmbeddingModel;
+
+import java.io.BufferedReader;
+import java.io.InputStreamReader;
+import java.io.Serializable;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.Map;
+import java.util.Scanner;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.regex.Pattern;
+
+/**
+ * This demo generates embeddings of movie plot summaries from the
+ * <a href="https://www.cs.cmu.edu/~ark/personas/">"CMU Movies Summary Corpus"</a>. The embeddings
+ * are stored in an IMap, along with movie metadata (title, release date).
+ * We then perform similarity search on user's input strings from console and list top 10 movies
+ * with most similar plot summaries.
+ *
+ * <p>Dependencies:</p>
+ * <ul>
+ *     <li>langchain4j for the <a href="https://huggingface.co/sentence-transformers/all-MiniLM-L6-v2">all-miniLM-L6-v2</a>
+ *     model that transforms text to 384 dimensional dense vector</li>
+ * </ul>
+ */
+public class TextSimilaritySearchImap {
+
+    static Map<String, String> movieIdToPlotSummary = new ConcurrentHashMap<>();
+    static Map<String, MovieMetadata> movieIdToMeta = new ConcurrentHashMap<>();
+
+    public static void main(String[] args) throws Exception {
+        // instantiate embedding model
+        EmbeddingModel embeddingModel = new AllMiniLmL6V2EmbeddingModel();
+
+        // start hazelcast
+        IMap<String, MovieMetadata> movies = startHzAndGetMovieImap();
+        // uncomment following line if you have a running cluster and want to execute this demo with a Hazelcast client
+//         IMap<String, MovieMetadata> movies = startHzClientAndGetMovieImap();
+        // index not necessary for a small IMap
+//        movies.addIndex(IndexType.SORTED, "releaseDate");
+
+        // ingest
+        loadSummaries(movies);
+        generateEmbeddings(movies, embeddingModel);
+
+        // read user input from console and search matches
+        try (BufferedReader consoleReader = new BufferedReader(new InputStreamReader(System.in))) {
+            while (true) {
+                // read user input
+                System.out.println("Enter release date pattern followed by search string:");
+                String line = consoleReader.readLine();
+                var parts = line.split(" ", 2);
+                if (parts.length != 2) {
+                    System.out.println("Invalid search string");
+                    continue;
+                }
+
+                String yearPattern = parts[0];
+                String userInput = parts[1];
+
+                // transform user input to vector
+                float[] query = embeddingModel.embed(userInput).content().vector();
+
+                // find & output top 10 similar matches of plot summary to given text
```

Comment:
maybe we could do 2 in 1 (via some flag etc), so code wouldn't be commented?

Reply:
ok, that will be probably less confusing: https://github.com/hazelcast/hazelcast-code-samples/pull/710/commits/a82b3b5e44e968a2793b991e982a9dd3111df44d